src/source/CurlApiCallbacks.c

```diff
@@ -1034,31 +1030,86 @@ STATUS createStreamCachingCurl(UINT64 customData, PCHAR deviceName, PCHAR stream
 {
     ENTERS();
 
-    UNUSED_PARAM(deviceName);
-    UNUSED_PARAM(contentType);
-    UNUSED_PARAM(kmsKeyId);
-    UNUSED_PARAM(retentionPeriod);
-
     STATUS retStatus = STATUS_SUCCESS;
     STREAM_HANDLE streamHandle;
-
     PCurlApiCallbacks pCurlApiCallbacks = (PCurlApiCallbacks) customData;
     PCallbacksProvider pCallbacksProvider = NULL;
+    BOOL endpointsLocked = FALSE, emulateApiCall = TRUE;
+    UINT64 curTime, value;
+    PEndpointTracker pEndpointTracker = NULL;
 
     CHK(pCurlApiCallbacks != NULL && pCurlApiCallbacks->pCallbacksProvider != NULL && pServiceCallContext != NULL, STATUS_INVALID_ARG);
     pCallbacksProvider = pCurlApiCallbacks->pCallbacksProvider;
 
-    DLOGV("No-op CreateStream API call");
-
     streamHandle = (STREAM_HANDLE) pServiceCallContext->customData;
 
+    // We check whether we have already made the call by checking
+    // for the presence of the endpoint in the cache.
+    switch(pCurlApiCallbacks->cacheType) {

```

Comment:
Similar pattern for all of the caching APIs. We check the caching logic first. If none then we just run the main function. If endpoint only (with the exception of the get endpoint API call) we emulate always. If it's cache all API calls, we look at the pre-cached endpoint and it's TTL. If all valid then we will emulate the calls, meaning that we had run these before.